pipit/trace.py

```diff
@@ -127,3 +127,159 @@ def comm_matrix(self, output="size"):
+            ]
+
+            for process in set(enter_leave_df["Process"]):
+                curr_process_df = enter_leave_df.loc[
+                    enter_leave_df["Process"] == process
+                ]
+                for thread in set(curr_process_df["Thread"]):
+                    # filter by both process and thread
+                    filtered_df = curr_process_df.loc[
+                        curr_process_df["Thread"] == thread
+                    ]
+
+                    stack = []
+
+                    """
+                    Note:
+                    The reason that we are creating lists that are copies of the
+                    dataframecolumns below and iterating over those instead of using
+                    pandas iterrows is due to an observed improvement in performance
+                    when using lists.
+                    """
+
+                    event_types = list(filtered_df["Event Type"])
+                    df_indices, timestamps = list(filtered_df.index), list(
+                        filtered_df["Timestamp (ns)"]
+                    )
+
+                    # Iterate through all events of filtered DataFrame
+                    for i in range(len(filtered_df)):
+                        curr_df_index, curr_timestamp, evt_type = (
+                            df_indices[i],
+                            timestamps[i],
+                            event_types[i],
+                        )
+
+                        if evt_type == "Enter":
+                            # Add current dataframe index and timestamp to stack
+                            stack.append((curr_df_index, curr_timestamp))
+                        else:
+                            # Pop corresponding enter event's dataframe index
+                            # and timestamp
+                            enter_df_index, enter_timestamp = stack.pop()
+
+                            # Fill in the lists with the matching values
+                            matching_indices[enter_df_index] = curr_df_index
+                            matching_indices[curr_df_index] = enter_df_index
+
+                            matching_times[enter_df_index] = curr_timestamp
+                            matching_times[curr_df_index] = enter_timestamp
+
+            self.events["Matching Index"] = matching_indices
+            self.events["Matching Timestamp"] = matching_times
+
+    def __gen_calling_relationships(self):
+        """
+        Three columns to be added to dataframe:
+        "Depth", "Parent", and "Children"
+
+        Depth is level in the call tree starting from 0.
+        Parent is the dataframe index of a row's parent event.
+        Children is a list of dataframe indices of a row's children events.
+        """
+
+        if "Children" not in self.events.columns:
+            children = [None] * len(self.events)
+            depth, parent = [float("nan")] * len(self.events), [float("nan")] * len(
```

Comment:
Shouldn't depth be an integer? same for parent

Reply:
will change this